contrib/exporters/secadvisor/mod/acc_rej_action.go

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (C) 2019 IBM, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy ofthe License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specificlanguage governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package mod
+
+import (
+	"time"
+
+	"github.com/skydive-project/skydive/flow"
+	//"github.com/skydive-project/skydive/logging"
+	"github.com/spf13/viper"
+)
+
+type FlowInfoStatus int32
+
+const (
+	FlowInfo_ONE_ENTRY   FlowInfoStatus = 1
+	FlowInfo_TWO_ENTRIES FlowInfoStatus = 2
+	FlowInfo_OLD_FLOW    FlowInfoStatus = 3
+)
+
+// time for cleanup of old connections
+const timeoutPeriod = 20
+
+type accRejFlowInfo struct {
+	Status                   FlowInfoStatus
+	UUID1                    string
+	UUID2                    string
+	MetricReported           flow.FlowMetric
+	LastUpdateMetricReported flow.FlowMetric // FIXME - we might not need this
+	Saw1                     bool
+	Saw2                     bool
+	Flow1                    *flow.Flow
+	Flow2                    *flow.Flow
+}
+
+type accRejAction struct {
+	// table of saved flows
+	accRejState map[string]*accRejFlowInfo
+	timeOfLastCleanup	int64
+}
+
+// NewAccRejAction - creates data structures needed for vpc Accept/Reject based on measurements flows at 2 points
+func NewAccRejAction(cfg *viper.Viper) (interface{}, error) {
+	// allocate the table to store flow state
+	accRejState := make(map[string]*accRejFlowInfo)
+	now := time.Now()
+	secs := now.Unix()
+	accRej := &accRejAction{
+		accRejState:  accRejState,
+		timeOfLastCleanup: secs,
+	}
+	return accRej, nil
+}
+
+// InitAccRejAction - initialize internal data structure for this round
+func (t *accRejAction) InitAccRejAction() {
+	// reset the entries in the table for this round
+	for _, ars := range t.accRejState {
+		ars.Saw1 = false
+		ars.Saw2 = false
+		ars.Status = FlowInfo_OLD_FLOW
+	}
+}
+
+// AccumulateAccRejInfo - go over all the flows and fill in our internal data structure with relevant info
+func (t *accRejAction) AccumulateAccRejInfo(fl []*flow.Flow) {
+	for _, f := range fl {
+		L3TrackingID := f.GetL3TrackingID()
```

Comment:
what about non-L3 flows? they should probably be transferred "as-is"

Reply:
you are counting on getting two flows with the same `f.L3TrackingID` from different `f.Layer.ID` but you must check that the `f.Layer.ID` of the two flows are not identical.

Reply:
The packet will already have either passed through or not. The question is only about the flow reporting. Is there flow information for a non-L3 packet? Does it matter if we do not report it? Let's discuss.

Reply:
there is a TrackingID field please check if that can serve non L3 traffic

Reply:
Apparently the L3TrackingID is also OK for non-L3 flows. Perhaps it just uses a port == 0 since the info is not available. I ran a simple test of ping and it picks it up OK. Not sure what will happen if we have conflicting non-L3 flows. Will check the TrackingID field.